### PR TITLE
[cxx-interop] Add support for benchmarks with C++ interop.

### PR DESCRIPTION
**Pull request by zoecarver**
* Adds support for benchmarks that use C++ modules.
* Adds "CreateObjects" benchmark that creates C++ objects.

Based on #32715.

### CODE REVIEW

File: benchmark/cxx-source/CreateObjects.swift
Comment (MForster) [inline benchmark/cxx-source/CreateObjects.swift:22]: > A benchmark should only be tagged "cpubench" after a full performance investigation of the benchmark has been completed to determine that it is a good representation of future Swift performance.

https://github.com/apple/swift/blob/de3af89539b9812f38a44730d7b468c2651eb10d/benchmark/utils/TestsUtils.swift#L64,L66

File: benchmark/cmake/modules/AddSwiftBenchmarkSuite.cmake
Comment (MForster) [inline benchmark/cmake/modules/AddSwiftBenchmarkSuite.cmake:589]: Why does the main module need this?

**Review by MForster — CHANGES_REQUESTED** [review]
This seems to contain a lot of unrelated changes...

File: benchmark/cxx-source/CreateObjects.swift
Comment (zoecarver) [reply benchmark/cxx-source/CreateObjects.swift:22]: Oh, my bad. I'll remove it.

**Comment by zoecarver** [comment]
> This seems to contain a lot of unrelated changes...

Yes, this PR is based on #32715 so all the changes from that PR are here too (just so it compiles).

File: benchmark/cmake/modules/AddSwiftBenchmarkSuite.cmake
Comment (zoecarver) [reply benchmark/cmake/modules/AddSwiftBenchmarkSuite.cmake:589]: Because otherwise when compiling `main.swift` the compiler complains about missing the C++ includes:
```
/Users/zoe/Developer/swift-source/swift/benchmark/utils/main.swift:51:8: error: missing required module 'CxxTypes'
import CreateObjects
```

**Review by MForster — APPROVED** [review]
(after the base pull request has been merged)

**Comment by zoecarver** [comment]
@MForster sounds good. Thanks!

File: benchmark/utils/CxxTests/cxx-types.h
Comment (gribozavr) [inline benchmark/utils/CxxTests/cxx-types.h:1]: I'd suggest to imitate what we do in the `test/Interop` directory, and create a separate supporting header for each benchmark; otherwise it would be very difficult to untangle supporting code that gets reused between benchmarks that was not actually meant to be reused.

File: lib/IRGen/IRGenSIL.cpp
Comment (gribozavr) [inline lib/IRGen/IRGenSIL.cpp:821]: Is there a separate PR for these changes to the compiler?

**Review by gribozavr — COMMENTED** [review]
@zoecarver This PR is adding `benchmark/.swiftpm/xcode/package.xcworkspace/xcuserdata/zoe.xcuserdatad/UserInterfaceState.xcuserstate`, I think that's not intentional.

File: benchmark/utils/CxxTests/cxx-types.h
Comment (zoecarver) [reply benchmark/utils/CxxTests/cxx-types.h:1]: Hmm, good idea. I'll do that.

File: lib/IRGen/IRGenSIL.cpp
Comment (zoecarver) [reply lib/IRGen/IRGenSIL.cpp:821]: No, there's not. I'll make one.

**Comment by zoecarver** [comment]
> This PR is adding benchmark/.swiftpm/xcode/package.xcworkspace/xcuserdata/zoe.xcuserdatad/UserInterfaceState.xcuserstate, I think that's not intentional.

Oops, you're right. Xcode adds that when I open the `benchmarks/` directory. Forgot to remove it last time I committed.

File: lib/IRGen/IRGenSIL.cpp
Comment (zoecarver) [reply lib/IRGen/IRGenSIL.cpp:821]: Here it is: #32815.

File: benchmark/utils/CxxTests/CreateObjects.h
Comment (gribozavr) [inline benchmark/utils/CxxTests/CreateObjects.h:1]: "CxxIntWrapper" b/c the int is not on the heap?

File: benchmark/utils/CxxTests/CreateObjects.h
Comment (gribozavr) [inline benchmark/utils/CxxTests/CreateObjects.h:2]: In the standard library, we'd usually call the member "value" in similar situations.

**Comment by gribozavr** [comment]
LGTM! Please rebase and merge after #32815 goes in.

**Review by gribozavr — APPROVED** [review]

File: benchmark/utils/CxxTests/CreateObjects.h
Comment (zoecarver) [reply benchmark/utils/CxxTests/CreateObjects.h:1]: SGTM.

**Comment by zoecarver** [comment]
@swift-ci  please test and merge

**Comment by zoecarver** [comment]
@swift-ci please smoke test and merge

**Comment by zoecarver** [comment]
@swift-ci please smoke test and merge

**Comment by zoecarver** [comment]
@swift-ci please test Linux platform.

**Comment by zoecarver** [comment]
@swift-ci please test Linux platform

**Comment by zoecarver** [comment]
@swift-ci please benchmark

**Comment by zoecarver** [comment]
@swift-ci please test windows platform.

**Comment by zoecarver** [comment]
@swift-ci please benchmark

**Comment by zoecarver** [comment]
@swift-ci please benchmark.

**Comment by zoecarver** [comment]
@swift-ci please benchmark

**Comment by zoecarver** [comment]
@swift-ci please benchmark

**Comment by zoecarver** [comment]
@swift-ci please benchmark

**Comment by zoecarver** [comment]
@swift-ci please smoke test

File: benchmark/cxx-source/CreateObjects.swift
Comment (eeckstein) [inline benchmark/cxx-source/CreateObjects.swift:25]: I assume this benchmark is just there to validate that C++ support is compiling fine.
I think it makes sense to disable this benchmark from the regular run, because it's not doing anything interesting.
Can you add the ".skip" benchmark tag?

Comment (zoecarver) [reply benchmark/cxx-source/CreateObjects.swift:25]: Actually, I'd like to test the performance of creating C++ objects. For example, I expect that this will be much slower after #30630 (when we actually have to call the constructor instead of just allocating an empty object).

Comment (eeckstein) [reply benchmark/cxx-source/CreateObjects.swift:25]: I see. Makes sense

**Comment by zoecarver** [comment]
@swift-ci please smoke test Linux platform.

**Comment by zoecarver** [comment]
@swift-ci please smoke test Linux platform

**Comment by zoecarver** [comment]
@swift-ci please smoke test Linux platform

**Comment by zoecarver** [comment]
@swift-ci please smoke test Linux platform

**Comment by zoecarver** [comment]
The build failure appears to be unrelated.

**Comment by zoecarver** [comment]
@swift-ci please smoke test Linux platform.

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Comment by zoecarver** [comment]
@swift-ci please smoke test Linux platform.

**Comment by zoecarver** [comment]
Failed because of an unrelated issue. 

@swift-ci please smoke test Linux platform.

**Comment by MaxDesiatov** [comment]
https://github.com/apple/swift-corelibs-foundation/pull/2895
@swift-ci please smoke test

**Comment by zoecarver** [comment]
@swift-ci please test.

**Comment by zoecarver** [comment]
@swift-ci please test.

**Pull request merged**
